### ISSUE TITLE
Feature Request:  Seamless stream and/or async iterable support

### ISSUE THREAD

**Issue by ronag** [jasnell/piscina#110]
Would be nice to be able to provide a stream and/or async iterable (of buffers) as argument and have is seamlessly (using transferable) accessible in the worker.

e.g.

```js
const handle = await fs.open(dst)
try {
  for await (const buf of worker.runTask(fs.createReadStream(src))) {
     await handle.write(buf)
  }
} finally {
  await handle.close()
}
```

```js
module.exports = async function * (source) {
  for await (const buf of source) {
    yield veryExpensiveProcessing(buf)
  }
}
```

**Comment by ronag** [comment]
maybe duplicate of https://github.com/piscinajs/piscina/issues/108

**Comment by ronag** [comment]
@mcollina This would make it possible to do e.g.

```js
await pipeline(piscina.move(src), piscina.transform(workerPath), dst)
```

**Comment by jasnell** [comment]
Unfortunately we can't make streams *seamlessly* transferable/cloneable using move. The only objects that can be made transferable are core objects that extend from BaseObject. It would be possible to create a Transform implementation whose _transform dispatched to Piscina and continued on after each task was done. So something like..


```js
const piscina = new Piscina({ /** ... **/ });
const workerTransform = new MyTransformer(piscina);
await pipeline(src, workerTransform, dst);
```

I have been thinking about adding a new type of Stream in core that complements MessagePort to implement a highly efficient cross-thread stream model. Basically something like:

```js
const { MessageDuplex } = require('worker_threads')

const mc = new MessageChannel();

mc.port1.onmessage = async ({data}) => {
  for await (const chunk of data)
    data.write(transformIt(chunk));
};

const d = new MessageDuplex();
d.on('data', console.log);

mc.port2.postMessage(d);

d.write('hello');
d.end('there');
```

The idea here is that `MessageDuplex` would use highly efficient signaling and no-copy mechanisms to exchange data with the paired clone. It would get very close to what you're looking for here.

bottom line is that the limitations of cloning/transfering over MessagePort definitely impose some restrictions.

**Comment by jasnell** [comment]
Regarding this pattern.... I'm not sure this (an async generator as a worker) is a pattern that could work with the worker pool. Retaining the generator state would be problematic. Not necessarily impossible but extremely complicated. Will have to think about how we could do it and what the semantics would actually be.

```js
module.exports = async function * (source) {
  for await (const buf of source) {
    yield veryExpensiveProcessing(buf)
  }
}
```

**Comment by ronag** [comment]
> I have been thinking about adding a new type of Stream in core that complements MessagePort to implement a highly efficient cross-thread stream model. Basically something like:

I think all of that could be hidden away in a `piscina.pipeline()` helper, no?

**Comment by jasnell** [comment]
Possibly, but we still hit up against limitations on MessagePort. Lemme stew over it a bit

**Comment by jasnell** [comment]
https://github.com/piscinajs/piscina/pull/170 illustrates using transferable whatwg readable and writable streams to an individual worker. It doesn't quite do what this issue is suggesting but it definitely gets a bit closer!

**Comment by metcoder95** [comment]
I've been dragging my head around it, and although wouldn't be the most performant, I believe we can achieve some sort of bare basic support for `streams/iterators` in some way.

Right now I'm thinking about the API and I imagine two things:

1. ~~`piscina.run` supports inputs a streams~~
  - ~~We can use specific `MessagePort` with a ring buffer to transfer the stream chunks (we can even cache the buffer for further transfers with some heuristics) and reconstruct the other end with a stream that reads from the buffer. The goal is to also support backpressure from the worker to the pool.~~
  - ~~If the worker returns an iterator/stream, we can either use the same buffer (if the input of the task was a stream) or we can create a new ring buffer that does the same job (same or different heuristics might apply)~~

2. `piscina.pipeline` ~~can benefit for the exact same `piscina.run` implementations with slightly adjustments.~~ can actually return a `RedeableStream` and expect either an AsyncIterator/Iterator or a Stream in return.
